extensions/blocks/premium-content/_inc/subscription-service/class-token-subscription-service.php

```diff
@@ -0,0 +1,254 @@
+				$is_valid_token = false;
+			}
+		}
+
+		if ( $is_valid_token ) {
+			$subscriptions = (array) $payload['subscriptions'];
+		} elseif ( is_user_logged_in() ) {
+			// If there is no token, but the user is logged in, get current subscriptions and determine if the user has
+			// a valid subscription to match the plan ID.
+			$subscriptions = apply_filters( 'earn_get_user_subscriptions_for_site_id', array(), wp_get_current_user()->ID, $this->get_site_id() );
+			if ( empty( $subscriptions ) ) {
+				return false;
+			}
+			// format the subscriptions so that they can be validated.
+			$subscriptions = self::abbreviate_subscriptions( $subscriptions );
+		} else {
+			return false;
+		}
+
+		return $this->validate_subscriptions( $valid_plan_ids, $subscriptions );
+	}
+
+	/**
+	 * Decode the given token.
+	 *
+	 * @param string $token Token to decode.
+	 *
+	 * @return array|false
+	 */
+	public function decode_token( $token ) {
+		try {
+			$key = $this->get_key();
+			return $key ? (array) JWT::decode( $token, $key, array( 'HS256' ) ) : false;
+		} catch ( \Exception $exception ) {
+			// TODO: The SignatureInvalidException is in a different namespace from JWT, so if we want to catch that
+			// one specifically then we'll have to alias it.
+			$logstash = array(
+				'feature' => self::DECODE_EXCEPTION_FEATURE,
+				'message' => self::DECODE_EXCEPTION_MESSAGE,
+				'extra'   => wp_json_encode( compact( 'exception', 'token' ) ),
+			);
+         // phpcs:ignore ImportDetection.Imports.RequireImports.Symbol
+			log2logstash( $logstash );
+			return false;
+		}
+	}
+
+	/**
+	 * Get the key for decoding the auth token.
+	 *
+	 * @return string|false
+	 */
+	abstract public function get_key();
+
+	/**
+	 * Get the ID of the current site.
+	 *
+	 * @return int
+	 */
+	abstract public function get_site_id();
+
+	// phpcs:disable
+	/**
+	 * Get the URL to access the protected content.
+	 *
+	 * @param string $mode Access mode (either "subscribe" or "login").
+	 */
+	public function access_url( $mode = 'subscribe' ) {
+		global $wp;
+		$permalink = get_permalink();
+		if ( empty( $permalink ) ) {
+			$permalink = add_query_arg( $wp->query_vars, home_url( $wp->request ) );
+		}
+
+		$login_url = $this->get_rest_api_token_url( $this->get_site_id(), $permalink );
+		return $login_url;
+	}
+	// phpcs:enable
+
+	/**
+	 * Get the token stored in the auth cookie.
+	 *
+	 * @return ?string
+	 */
+	private function token_from_cookie() {
+		if ( isset( $_COOKIE[ self::JWT_AUTH_TOKEN_COOKIE_NAME ] ) ) {
+			// phpcs:ignore WordPress.Security.ValidatedSanitizedInput.MissingUnslash, WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
+			return $_COOKIE[ self::JWT_AUTH_TOKEN_COOKIE_NAME ];
+		}
+	}
+
+	/**
+	 * Store the auth cookie.
+	 *
+	 * @param  string $token Auth token.
+	 * @return void
+	 */
+	private function set_token_cookie( $token ) {
+		if ( ! empty( $token ) ) {
+			setcookie( self::JWT_AUTH_TOKEN_COOKIE_NAME, $token, 0, '/' );
+		}
+	}
```

Comment:
This seems to be causing some issues on my end. When I try to login I am redirected to the page instead, with the following errors:

```
Warning: Cannot modify header information - headers already sent by (output started at /wp-content/themes/twentytwentyone/header.php:15) in /wp-content/plugins/jetpack-dev/extensions/blocks/premium-content/_inc/subscription-service/class-token-subscription-service.php on line 167

Warning: Cannot modify header information - headers already sent by (output started at /wp-content/themes/twentytwentyone/header.php:15) in /wp-content/plugins/jetpack-dev/extensions/blocks/premium-content/_inc/subscription-service/class-token-subscription-service.php on line 167
```

In an incognito window (where third-party cookies were disabled), I am not given the opportunity to log in, as mentioned above. In a new browser, I can log in. Since I did not find how to subscribe I tried logging in with an account that was not subscribed to the site. I then got redirected to the site, with no notice about a successful or unsuccessful connection, and with the notices above.

Reply:
I tried this on a WordPress.com site in an incognito window and everything went fine. I didn't see the above errors. Subscribe and login buttons are present.